### ISSUE TITLE
linux debain 2.6.32-5-amd64, shell script be compress to  ELF 32-bit.

### ISSUE THREAD

**Issue by zsmsky** [upx/upx#53]
how  can I do? 
The following is my operation process！

root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# uname -r
2.6.32-5-amd64
root@debian:/home/upx-3.92-amd64_linux# ll
total 568
-rw-r--r-- 1 root root   1750 Dec 12 00:52 BUGS
-rw-r--r-- 1 root root  18092 Dec 12 00:52 COPYING
-rw-r--r-- 1 root root   5448 Dec 12 00:52 LICENSE
-rw-r--r-- 1 root root  22006 Dec 12 00:52 NEWS
-rw-r--r-- 1 root root   4587 Dec 12 00:52 README
-rw-r--r-- 1 root root    800 Dec 12 00:52 README.1ST
-rwxr-xr-x 1 root root  17078 Jan 13 16:06 test.sh
-rw-r--r-- 1 root root   2198 Dec 12 00:52 THANKS
-rwxr-xr-x 1 root root 354984 Dec 12 00:52 upx
-rw-r--r-- 1 root root  43059 Dec 12 00:52 upx.1
-rw-r--r-- 1 root root  37211 Dec 12 00:52 upx.doc
-rw-r--r-- 1 root root  39111 Dec 12 00:52 upx.html
root@debian:/home/upx-3.92-amd64_linux# ./upx -o test.run test.sh
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2016
UPX 3.92        Markus Oberhumer, Laszlo Molnar & John Reiser   Dec 11th 2016

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
     17078 ->      6479   37.94%  linux.sh/i386  test.run                      

Packed 1 file.
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# ./test.run 
Segmentation fault
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# file test.run 
test.run: ELF 32-bit LSB executable, Intel 80386, version 1 (SYSV), statically linked, corrupted section header size
root@debian:/home/upx-3.92-amd64_linux# 
root@debian:/home/upx-3.92-amd64_linux# 

ps:
I make sure that the script(test.sh) is executed correctly

## How can we reproduce the issue?

1.
2.
3.
4.

## Please tell us details about your environment.

* UPX version used (`upx --version`):
* Operating System and version:
* CPU architecture:

**Comment by jreiser** [comment]
I do not see an explicit problem or question in the Opening Comment of this Issue. I do detect the possibility of a problem with communicating in English. Therefore I comment on a couple items that might be issues.

When UPX compresses a shell script then the output format always is linux.sh/i386.  The output format must be a binary format in order for compression to work reasonably, so the only choice is which binary format.  UPX chooses ELF for i386.

The command "file test.run" says "corrupted section header size" because the 'file' command is simplistic.  Using "readelf --headers test.run" to display the headers:
```
  Start of section headers:          0 (bytes into file)
  Flags:                             0x0
  Size of this header:               52 (bytes)
  Size of program headers:           32 (bytes)
  Number of program headers:         2
  Size of section headers:           0 (bytes)
  Number of section headers:         0
  Section header string table index: 0
```
Notice that Start, Size, and Number of section headers are all 0, and so is Section header string table index.  These fields of the Elf32_Ehdr are irrelevant for execution; the operating system ignores them.  Supplying "dummy" contents would take space, therefore UPX does not.  However, 'file' does not like all those zeroes.

**Comment by zsmsky** [comment]
Thank you very much for your reply！

But, when I run the 'test.run' , it  appears segmentation fault.

Why? What should I do?

How can I compresses a shell script, then run on the 64bit system, but 32bit system!

root@debian:/home/upx-3.92-amd64_linux#
root@debian:/home/upx-3.92-amd64_linux# ./test.run
Segmentation fault
root@debian:/home/upx-3.92-amd64_linux#
root@debian:/home/upx-3.92-amd64_linux#

**Comment by markus-oberhumer** [comment]
Compressing a shell script should result in a statically linked i386-elf32 binary that should work on
all 32-bit and 64-bit kernels.

So we will need more info about your problem.

**Comment by jreiser** [comment]
The result is a statically-linked i386-elf32 binary.  Executing that binary attempts a "user-mode execve()" of "/bin/sh" which fails on x86_64 because there is no way for user-mode code to switch from i386 to x86_64.  Actually there is a bug in the UPX stub for linux.sh/i386 because /bin/sh having .e_machine != EM_386 was not detected.

I have a fix for the bug: detect that "/bin/sh" is non-EM_386, then do an actual execve() system call.  Integrating the fix is hung-up with the fix for #43.

**Comment by jreiser** [comment]
Fixed in devel branch commit 564f630a25440bd8adcc5d18118bd8e0d00b9df4 .
